core/imt/inc/ROOT/RTaskArena.hxx

```diff
@@ -0,0 +1,79 @@
+// @(#)root/thread:$Id$
+// // Author: Xavier Valls Pla   08/05/20
+//
+/*************************************************************************
+ * Copyright (C) 1995-2020, Rene Brun and Fons Rademakers.               *
+ * All rights reserved.                                                  *
+ *                                                                       *
+ * For the licensing terms see $ROOTSYS/LICENSE.                         *
+ * For the list of contributors see $ROOTSYS/README/CREDITS.             *
+ *************************************************************************/
+
+//////////////////////////////////////////////////////////////////////////
+//                                                                      //
+// RTaskArena                                                           //
+//                                                                      //
+// This file implements the method to initialize and retrieve ROOT's    //
+// global task arena, together with a method to check for active        //
+// CPU bandwith control, and a class to wrap the tbb task arena with    //
+// the purpose of keeping tbb off the installed headers                 //
+//                                                                      //
+//////////////////////////////////////////////////////////////////////////
+
+#ifndef ROOT_RTaskArena
+#define ROOT_RTaskArena
+
+#include "RConfigure.h"
+#include <memory>
+
+// exclude in case ROOT does not have IMT support
+#ifndef R__USE_IMT
+// No need to error out for dictionaries.
+# if !defined(__ROOTCLING__) && !defined(G__DICTIONARY)
+#  error "Cannot use ROOT::Internal::RTaskArenaWrapper without defining R__USE_IMT."
+# endif
+#else
+
+/// tbb::task_arena is an alias of tbb::interface7::task_arena, which doesn't allow
+/// to forward declare tbb::task_arena without forward declaring tbb::interface7
+namespace tbb{
+namespace interface7{class task_arena;}
+using task_arena = interface7::task_arena;
+}
+
+namespace ROOT {
+namespace Internal {
+
+////////////////////////////////////////////////////////////////////////////////
+/// Wrapper for tbb::task_arena.
+///
+/// Necessary in order to keep tbb away from ROOT headers.
+/// This class is thought out to be used as a singleton.
```

Comment:
So should it be a singleton, then? I know that the singleton is out of fashion, so maybe not mention "Singleton" here and say how to instantiate one?
I guess it shouldn't be a singleton. Maybe people want to have their second task arena for their framework, and probably that would work.

Reply:
The idea was to have a single task arena. If someone wants a second task arena, they can manipulate tbb directly.

Reply:
@hageboeck `RTaskArenaWrapper` is a singleton, not `tbb::task_arena`.

btw about actually making this a singleton see my comment about changing RTaskArenaWrapper's constructor to private

Reply:
That was my point. It's not a singleton, yet, but it could be. I'm not sure if it *need* be, though.

Reply:
yes, the whole design revolves around ROOT only ever having zero or one RTaskArenaWrappers in flight